examples/nlp/data_parallel_training_with_keras_nlp.py

```diff
@@ -0,0 +1,211 @@
+    .filter(lambda x: tf.strings.length(x) > 100)
+    .batch(PRETRAINING_BATCH_SIZE)
+)
+wiki_val_ds = (
+    tf.data.TextLineDataset(wiki_dir + "wiki.valid.token")
+    .filter(lambda x: tf.strings.length(x) > 100)
+    .batch(PRETRAINING_BATCH_SIZE)
+)
+wiki_test_ds = (
+    tf.data.TextLineDataset(wiki_dir + "wiki.test.tokens")
+    .filter(lambda x: tf.strings.length(x) > 100)
+    .batch(PRETRAINING_BATCH_SIZE)
+)
+
+"""
+In the above code, we download the wikitext-2 dataset and extract it. Then, we define 
+three datasets: wiki_train_ds, wiki_val_ds, and wiki_test_ds. These datasets are 
+filtered to remove short lines and are batched for efficient training.
+
+Let's define a function for decaying the learning rate. You can define any decay function you need.
+We also define a callback for printing the learning rate at the end of each epoch.
+"""
+
+
+def decay(epoch):
+    if epoch < 3:
+        return 1e-3
+    elif epoch >= 3 and epoch < 7:
+        return 1e-4
+    else:
+        return 1e-5
+
+
+class PrintLR(tf.keras.callbacks.Callback):
+    def on_epoch_end(self, epoch, logs=None):
+        print(
+            "\nLearning rate for epoch {} is {}".format(
+                epoch + 1, model_dist.optimizer.lr.numpy()
+            )
+        )
+
+
+"""
+Let's also make a callback to TensorBoard, this will enable visualization of different metrics
+while we train the model in later part of this tutorial We put all the callbacks together as follows:
+"""
+callbacks = [
+    tf.keras.callbacks.TensorBoard(log_dir="./logs"),
+    tf.keras.callbacks.LearningRateScheduler(decay),
+    PrintLR(),
+]
+
+
+print(tf.config.list_physical_devices("GPU"))
+
+"""
+To do single-host, multi-device synchronous training with a Keras model, you would use
+the [`tf.distribute.MirroredStrategy` API](
+    https://www.tensorflow.org/api_docs/python/tf/distribute/MirroredStrategy).
+Here's how it works:
+
+- Instantiate a `MirroredStrategy`, optionally configuring which specific devices you
+want to use (by default the strategy will use all GPUs available).
+- Use the strategy object to open a scope, and within this scope, create all the Keras
+objects you need that contain variables. Typically, that means **creating & compiling the
+model** inside the distribution scope.
+- Train the model via `fit()` as usual.
+"""
+
+strategy = tf.distribute.MirroredStrategy()
+print("Number of devices: {}".format(strategy.num_replicas_in_sync))
+
+"""
+With the datasets prepared, we now initialize and compile our model and optimizer within
+the `strategy.scope()`:
+"""
+with strategy.scope():
+    # Everything that creates variables should be under the strategy scope.
+    # In general this is only model construction & `compile()`.
+    model_dist = keras_nlp.models.BertMaskedLM.from_preset("bert_tiny_en_uncased")
+    model_dist.compile(
+        loss=keras.losses.SparseCategoricalCrossentropy(from_logits=True),
+        optimizer=keras.optimizers.Adam(5e-5),
```

Comment:
This learning rate is contradicting with the callback defined above. Basically you start with 5e-5, then at the end of first epoch it's bumped up to 1e-3.